presto-main/src/main/java/com/facebook/presto/metadata/ParametricScalar.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.metadata;
+
+import com.facebook.presto.operator.scalar.ScalarFunctionImplementation;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeManager;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.metadata.FunctionType.SCALAR;
+import static java.util.Objects.requireNonNull;
+
+public abstract class ParametricScalar
+        implements ParametricFunction
+{
+    private final Signature signature;
+
+    protected ParametricScalar(String name, List<TypeParameter> typeParameters, String returnType, List<String> argumentTypes)
+    {
+        this(name, typeParameters, returnType, argumentTypes, false);
+    }
+
+    protected ParametricScalar(String name, List<TypeParameter> typeParameters, String returnType, List<String> argumentTypes, boolean variableArity)
+    {
+        requireNonNull(name, "name is null");
+        requireNonNull(typeParameters, "typeParameters is null");
+        requireNonNull(returnType, "returnType is null");
+        requireNonNull(argumentTypes, "argumentTypes is null");
+        this.signature = new Signature(name, SCALAR, ImmutableList.copyOf(typeParameters), returnType, ImmutableList.copyOf(argumentTypes), variableArity);
+    }
+
+    @Override
+    public final Signature getSignature()
+    {
+        return signature;
+    }
+
+    public abstract ScalarFunctionImplementation specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry);
```

Comment:
@cberner Please have a look at our decimal work: https://github.com/sopel39/presto/commit/0a937b9ee07e1f40574a18b6a3b775ba0a77d036#diff-2d916d7fbaa4c1c57bd195f4eefa9c63 (specialize() function, subject to change). ScalarPolymorphicParametricFunction chooses function based on both argument types and return type (e.g. long decimal vs short decimal). In order to compute this we need type signatures for all parameters. Those are also required if we want to have some annotations for filtering that use literals. Such mechanism should rather be outside of FunctionRegistry. You mentioned that you are going to work on framework for annotations. Doesn't that conflict with @dain work on literals and annotation based type inference? Shouldn't that go first? With decimal work we have some advanced cases what such annotation framework should support. My idea after meeting with @dain and @martint was that our team in Boston/Warsaw will propose annotation based function definitions as part of decimal work.